Employee configuration

lomamech opened the pull request:
feature(Employees Configuration for Payroll)  …
- Implement module and API for the Configuration of Employees for Payroll
- Implement component EmployeeConfigSelect
- Improve Payroll's Select Component ADD message when not configuration found
- Updata data.sql, schema.sql
- (test) Update Integration test for Create Payroll Configuration
- Update Employee Registration manage property employee locked
- Update and Implemente Integration test for Employee Configuration
- Improve data.sql for easy Integration Test
- Implement E2E test for employee configuration Management

Closes #2835

lomamech committed:
feature(Employees Configuration for Payroll)
- Implement module and API for the Configuration of Employees for Payroll

lomamech committed:
refactor(Employee Configuration)
- Implement component EmployeeConfigSelect
- Improve Payroll's Select Component ADD message when not configuration
  found
- Updata data.sql, schema.sql
- (test) Update Integration test for Create Payroll Configuration

lomamech committed:
refactor(Employee Configuration)
- Update Employee Registration manage property employee locked
- Update and Implemente Integration test for Employee Configuration
- Improve data.sql for easy Integration Test

lomamech committed:
test(Employee Configuration)
- Implement E2E test for employee configuration Management

lomamech committed:
improvememt(Employee Configuration)
- Sanitaze Code

lomamech committed:
Fix Error

lomamech committed:
Fix Error

lomamech committed:
Resolve conflict with Branch Master

jniles commented on client/src/js/components/bhEmployeeConfigSelect.js:
```diff
@@ -0,0 +1,65 @@
+    $ctrl.label = $ctrl.label || 'EMPLOYEE.CONFIGURATION';
+
+    // fired when an employee configuration has been selected
+    $ctrl.onSelectCallback = $ctrl.onSelectCallback || angular.noop;
```
You don't need to set this default, since you have it required in the component binding (`&`).

jniles commented on client/src/modules/employees/registration/employees.html:
```diff
@@ -156,6 +156,19 @@
 
 
+                  <!-- is locked ?  -->
+                  <div id="locked" class="form-group"
```
Do you need to put this in a `form-group`?  I think a plain checkbox will work:
```html
<div class="checkbox">
  <!-- html -->
</div>
```

jniles commented on client/src/modules/payroll/employee_configuration/configuration.html:
```diff
@@ -0,0 +1,45 @@
+      </div>
+
+      <div class="toolbar-item">
+        <button
```
Can you use `<bh-filter-toggle></bh-filter-toggle>` here?

jniles commented on client/src/modules/payroll/employee_configuration/modals/config.modal.html:
```diff
@@ -0,0 +1,53 @@
+          <label style="padding-left: 12%">
+            <div ng-if="!employee.locked">
+              <input type="checkbox" ng-model="employee.checked">
+              <span><strong>{{ employee.display_name }}</strong></span>
```
I don't think you need the `<span>` here.  Is it useful?

jniles commented on client/src/modules/payroll/employee_configuration/configuration.service.js:
```diff
@@ -0,0 +1,32 @@
+ * @extends PrototypeApiService
+ *
+ * @description
+ * Encapsulates common requests to the /weekend_config/ URL.
```
This should say `employee_config` instead of `weekend_config`.

jniles commented on client/src/modules/payroll/employee_configuration/configuration.service.js:
```diff
@@ -0,0 +1,32 @@
+  service.setEmployees = setEmployees;
+
+  // loads the configuration's Employee
+  function getEmployees(id) {
```
Can this be called `getEmployeeConfiguration()`?  I thought it loaded employees at first, but it looks like it is only for loading the configuration of a single employee.

jniles commented on client/src/modules/payroll/employee_configuration/modals/config.modal.js:
```diff
@@ -0,0 +1,77 @@
+  function submit(employeeConfigForm) {
+    if (employeeConfigForm.$invalid || employeeConfigForm.$pristine) { return 0; }
+
+    const employeesChecked = vm.employees.filter(employee => employee.checked)
```
This is awesome functional programming 👍

jniles commented on client/src/modules/templates/bhEmployeeConfigSelect.tmpl.html:
```diff
@@ -0,0 +1,32 @@
+<div ng-form="{{ $ctrl.name }}" bh-employee-config-select ng-model-options="{ updateOn: 'default' }">
+  <!-- This component will only be displayed if there is no EMPLOYEE tax configuration, this will prevent an empty selection list from appearing in cases where the institution or company is not using the EMPLOYEE tax -->
```
Nice comments 👍

jniles commented on server/models/schema.sql:
```diff
@@ -1982,4 +1984,27 @@ CREATE TABLE `config_rubric_item` (
+  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `label` text,
+  PRIMARY KEY (`id`)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```
This charset should be `utf8` if we want to support accents.

jniles commented on server/models/schema.sql:
```diff
@@ -1982,4 +1984,27 @@ CREATE TABLE `config_rubric_item` (
+  UNIQUE KEY  (`config_employee_id`, `employee_uuid`),
+  FOREIGN KEY (`config_employee_id`) REFERENCES `config_employee` (`id`),
+  FOREIGN KEY (`employee_uuid`) REFERENCES `employee` (`uuid`)
+) ENGINE=InnoDB DEFAULT CHARSET=latin1;
```
Same comment - this CHARSET should be `utf8`.

jniles commented on test/end-to-end/employees/registration.spec.js:
```diff
@@ -112,28 +112,4 @@ describe('Employees', () => {
   });
-
-  // FIXME: skip throws an error
-  // it.skip('edits an employee', () => {
```
💃

jniles reviewed (commented):
@lomamech great work.

I've provided some feedback.  Let me know what you think!

lomamech committed:
refactor(employeeConfiguration)
- Remove not necessary callback
- Remove form-group for checkbox element
-  Use bh-filter-toogle
- Sanitaze code
- Use getEmployeeConfiguration instead getEmployee
- Use charset utf8 instead latin1

lomamech committed:
Update size for abbr on project and rubric_payroll

jniles commented on client/src/js/components/bhEmployeeConfigSelect.js:
```diff
@@ -0,0 +1,62 @@
+    $ctrl.name = $ctrl.name || 'EmployeeConfigForm';
+
+
+    if (!angular.isDefined($ctrl.required)) {
```
👍

jniles reviewed (approved):
LGTM.  Changes approved.

bors r+
